pkg/alertmanager/alertmanager.go

```diff
@@ -575,3 +607,117 @@ func (t *tenantRateLimits) RateLimit() rate.Limit {
 func (t *tenantRateLimits) Burst() int {
 	return t.limits.NotificationBurstSize(t.tenant, t.integration)
 }
+
+var (
+	errTooManyAlerts = "too many alerts, limit: %d"
+	errAlertsTooBig  = "alerts too big, total size limit: %d bytes"
+)
+
+const (
+	insertFailureTooManyAlerts = "too_many_alerts"
+	insertFailureAlertsTooBig  = "alerts_too_big"
+)
+
+type alertsLimiter struct {
+	tenant string
+	limits Limits
+
+	failureCounter *prometheus.CounterVec
+
+	mx        sync.Mutex
+	sizes     map[model.Fingerprint]int
+	count     int
+	totalSize int
+}
+
+func newAlertsLimiter(tenant string, limits Limits, failureCounter *prometheus.CounterVec) *alertsLimiter {
+	return &alertsLimiter{
+		tenant:         tenant,
+		limits:         limits,
+		sizes:          map[model.Fingerprint]int{},
+		failureCounter: failureCounter,
+	}
+}
+
+func (a *alertsLimiter) PreStore(alert *types.Alert, existing bool) error {
+	if alert == nil {
+		return nil
+	}
+
+	countLimit := a.limits.AlertmanagerMaxAlertsCount(a.tenant)
+	sizeLimit := a.limits.AlertmanagerMaxAlertsSizeBytes(a.tenant)
+
+	newSize := alertSize(alert.Alert)
+
+	a.mx.Lock()
+	defer a.mx.Unlock()
+
+	if !existing {
+		if countLimit > 0 && (a.count+1) > countLimit {
+			a.failureCounter.WithLabelValues(insertFailureTooManyAlerts).Inc()
+			return fmt.Errorf(errTooManyAlerts, countLimit)
+		}
+
+		if sizeLimit > 0 && (a.totalSize+newSize) > sizeLimit {
+			a.failureCounter.WithLabelValues(insertFailureAlertsTooBig).Inc()
+			return fmt.Errorf(errAlertsTooBig, sizeLimit)
+		}
+	}
+
+	return nil
+}
+
+func (a *alertsLimiter) PostStore(alert *types.Alert, existing bool) {
+	if alert == nil {
+		return
+	}
+
+	newSize := alertSize(alert.Alert)
+
+	a.mx.Lock()
+	defer a.mx.Unlock()
+
+	fp := alert.Fingerprint()
```

Comment:
nit: you can do this before taking the lock to minimise lock time

Reply:
Thanks. Indeed, that was my plan, and I failed to follow it :) Let me fix it.